src/cc/BPF.cc

```diff
@@ -0,0 +1,236 @@
+      probe_event.c_str(), probe_event_desc.c_str(),
+      pid, cpu, group_fd,
+      cb, cb_cookie
+    );
+
+    if (!res)
+      throw std::runtime_error(
+        "Unable to attach " +
+        ((ret_probe ? "return " : "") + ("uprobe for binary " + binary_path)) +
+        " symbol " + symbol + " address " + std::to_string(symbol_addr) +
+        " using probe function " + probe_func
+      );
+
+    uprobes_[probe_event] = res;
+  }
+
+  void BPF::detach_kprobe(const std::string& kernel_func, bool ret_probe) {
+    std::string event = get_kprobe_event(kernel_func, ret_probe);
+    auto it = kprobes_.find(event);
+    if (it == kprobes_.end())
+      throw std::invalid_argument(
+        "No open " +
+        ((ret_probe ? "return " : "") + ("kprobe for " + kernel_func))
+      );
+    detach_kprobe_event(it->first, it->second);
+    kprobes_.erase(it);
+  }
+
+  void BPF::detach_uprobe(
+    const std::string& binary_path,
+    const std::string& symbol,
+    uint64_t symbol_addr,
+    bool ret_probe
+  ) {
+    auto sym = check_binary_symbol(binary_path, symbol, symbol_addr);
+    std::string event = get_uprobe_event(sym->module, sym->offset, ret_probe);
+    auto it = uprobes_.find(event);
+    if (it == uprobes_.end())
+      throw std::invalid_argument(
+        "No open " +
+        ((ret_probe ? "return " : "") + ("uprobe for binary " + binary_path)) +
+        " symbol " + symbol + " address " + std::to_string(symbol_addr)
+      );
+    detach_uprobe_event(it->first, it->second);
+    uprobes_.erase(it);
+  }
+
+  int BPF::load_func(const std::string& func_name, enum bpf_prog_type type) {
+    if (funcs_.find(func_name) != funcs_.end())
+      return funcs_[func_name];
+
+    uint8_t* func_start = bpf_module_->function_start(func_name);
+    if (!func_start)
+      return -1;
+    size_t func_size = bpf_module_->function_size(func_name);
+
+    int fd = bpf_prog_load(
+      type,
+      reinterpret_cast<struct bpf_insn*>(func_start),
+      func_size,
+      bpf_module_->license(),
+      bpf_module_->kern_version(),
+      nullptr, 0 // BPFModule will handle error printing
+    );
+
+    if (fd < 0)
+      return -1;
+    funcs_[func_name] = fd;
+    return fd;
+  }
+
+  std::shared_ptr<bcc_symbol> BPF::check_binary_symbol(
+    const std::string& binary_path,
+    const std::string& symbol,
+    uint64_t symbol_addr
+  ) {
+    auto sym = std::make_shared<bcc_symbol>();
+    int res = bcc_resolve_symname(
+      binary_path.c_str(), symbol.c_str(), symbol_addr, sym.get()
+    );
+    if (res < 0)
+      throw std::runtime_error(
+        "Unable to find offset for binary " + binary_path +
+        " symbol " + symbol + " address " + std::to_string(symbol_addr)
+      );
+    return sym;
+  }
+
+  std::string BPF::get_kprobe_event(
+    const std::string& kernel_func,
+    bool ret_probe
+  ) {
+    std::string res = ret_probe ? "r_" : "p_";
+    res += std::regex_replace(kernel_func, KPROBE_EV_NAME_REPLACE, "_");
```

Comment:
@4ast I think I know why this is not done in C side. The event description always look like `<TYPE PREFIX>/<EVENT_NAME> <RAW_EVENT_NAME>`, where EVENT_NAME is the formatted version but the <RAW_EVENT_NAME> is just the original string (kernel event name / binary path / etc.).

If I read correct, the event description doesn't actually matter so much. So I assume we could just to the replacement formatting for the entire thing? Or even better just construct the event description on C side?